lib/AssemblyUtil/FastaToAssembly.py

```diff
@@ -24,6 +32,29 @@
 def _upa(object_info):
     return f'{object_info[6]}/{object_info[0]}/{object_info[4]}'
 
+def _get_serialized_object_size(assembly_object):
+    arg_hash = {'params': assembly_object}
+    serialized = json.dumps(arg_hash, cls=_JSONObjectEncoder)
+    return sys.getsizeof(serialized)
+
+def _validate_workers_params(worker_count, var_name):
+    if worker_count is None:
+        raise ValueError(f"{var_name} is required")
+    if not worker_count.isdigit():
+        raise ValueError(f"{var_name} must be a digit")
+    if int(worker_count) < 1:
+        raise ValueError(f"{var_name} must be >= 1")
+    return int(worker_count)
+
+def _get_num_workers():
+    num_workers = os.environ.get(f"KBASE_SECURE_CONFIG_PARAM_{_NUM_WORKERS}", str(_DEFAULT_NUM_WORKERS))
+    max_num_workers = os.environ.get(f"KBASE_SECURE_CONFIG_PARAM_{_MAX_NUM_WORKERS}", str(_DEFAULT_MAX_NUM_WORKERS))
+    num_workers = _validate_workers_params(num_workers, _NUM_WORKERS)
+    max_num_workers = _validate_workers_params(max_num_workers, _MAX_NUM_WORKERS)
+    if num_workers > max_num_workers:
+        raise ValueError(f"Number of workers {num_workers} is greater than max number of workers {max_num_workers}")
```

Comment:
I don't understand the point of this. Whoever sets num workers sets max num workers as well, so they're never going to deliberately trigger this.

There should be 2 params - `threads_per_cpu` and `max_threads` and the worker count would be figured out via
```
workers = min(threads_per_cpu * cpu_count(), max_threads)
```

Reply:
Updated!

Reply:
I just realized by setting threads_per_cpu, now it's pretty hard to get an odd number like 5.

Reply:
Not if you set it to 2.5

Reply:
Also there should be sane defaults for the params, and it'd be helpful to print them out for debugging purposes

Reply:
Also also, threads / cpu could be < 1.

Reply:
> Also there should be sane defaults for the params

Are sane defaults refer to whatever ppl set in the catalog or what we initiated for them?

Reply:
The current code will error out if nothing is set in the catalog, so defaults can't be set there. If there's nothing in the catalog, the app should choose something reasonable and keep going

Reply:
Updated!

Reply:
* You can put the code and defaults that are currently in the impl method block in the class header or module header block. * Never use `eval` to parse external input. Why not use `float` or `int` like before?

Reply:
I can put defaults inside the class header. Where exactly is module header in the Impl file?

Reply:
Looking at the code some more - I would parse to numbers in the Impl file, but check values (e.g. > 0, > 1) in FastaToAssembly. FtA should be responsible for ensuring its inputs are ok, but not coercing them to the right type

Reply:
> Where exactly is module header block in the Impl file?

https://github.com/kbaseapps/AssemblyUtil/blob/develop/lib/AssemblyUtil/AssemblyUtilImpl.py#L2-L13

Reply:
I assume max_threads needs to be > 0 and can be decimal as well?

Reply:
Nope, need to be > 1

Reply:
Max threads doesn't make much sense as a float, so I'd say it has to be an int and > 0. 1 thread is fine

Reply:
Updated!

Reply:
: